cedar-policy-validator/src/schema.rs

```diff
@@ -207,6 +207,86 @@ impl TryFrom<json_schema::Fragment<RawName>> for ValidatorSchema {
 }
 
 impl ValidatorSchema {
+    /// Returns an iterator over every entity type that can be a principal for any action in this schema
+    pub fn principals(&self) -> impl Iterator<Item = &EntityType> {
+        self.action_ids
+            .values()
+            .flat_map(ValidatorActionId::principals)
+    }
+
+    /// Returns an iterator over every entity type that can be a resource for any action in this schema
+    pub fn resources(&self) -> impl Iterator<Item = &EntityType> {
+        self.action_ids
+            .values()
+            .flat_map(ValidatorActionId::resources)
+    }
+
+    /// Returns an iterator over every entity type that can be a principal for `action` in this schema
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if `action` is not found in the schema
+    pub fn principals_for_action(
+        &self,
+        action: &EntityUID,
+    ) -> Option<impl Iterator<Item = &EntityType>> {
+        self.action_ids
+            .get(action)
+            .map(ValidatorActionId::principals)
+    }
+
+    /// Returns an iterator over every entity type that can be a resource for `action` in this schema
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if `action` is not found in the schema
+    pub fn resources_for_action(
+        &self,
+        action: &EntityUID,
+    ) -> Option<impl Iterator<Item = &EntityType>> {
+        self.action_ids
+            .get(action)
+            .map(ValidatorActionId::resources)
+    }
+
+    /// Returns an iterator over all the entity types that can be a parent of `ty`
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if the `ty` is not found in the schema
+    pub fn parents<'a>(
```

Comment:
If I understand correctly from the implementation, this actually means (transitive) ancestors?  Should we name the method `ancestors()`?